rust/catalyst-types/src/id_uri/role_index.rs

```diff
@@ -12,54 +12,82 @@ use thiserror::Error;
 /// Role Index parsing error
 #[derive(Display, Error, Debug)]
 #[allow(clippy::module_name_repetitions)]
-pub enum RoleIndexError {
+pub enum RoleIdError {
+    /// Invalid parse
+    Parse(#[from] ParseIntError),
+
     /// Invalid Role Index
-    InvalidRole(#[from] ParseIntError),
+    InvalidRole(u8),
 }
 
 /// Project Catalyst User Role Index.
 ///
 /// <https://github.com/input-output-hk/catalyst-CIPs/blob/x509-catalyst-role-definitions/CIP-XXXX/README.md>
 #[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
-pub struct RoleIndex(u16);
-
-impl RoleIndex {
-    /// Delegated Representative
-    pub const DREP: RoleIndex = RoleIndex(1);
-    /// Proposer
-    pub const PROPOSER: RoleIndex = RoleIndex(3);
-    /// Default Role 0
-    pub const ROLE_0: RoleIndex = RoleIndex(0);
+#[repr(u8)]
+#[non_exhaustive]
+pub enum RoleId {
+    /// Can Vote and Comment on Proposals.
+    Role0 = 0,
+    /// Allows for voters to give their voting power to the dRep who then votes on their
+    /// behalf.
+    DelegatedRepresentative = 1,
+    /// Assigns a voters voting power to a registered dRep.
+    VoterDelegation = 2,
+    /// Allows the Propers signing key to also be used as a public encryption key.
+    Proposer = 3,
+}
 
-    /// Is the `RoleIndex` the default value (Role 0)
+impl RoleId {
+    /// Is the `RoleId` the default value [`Self::Role0`]
     #[must_use]
     pub fn is_default(self) -> bool {
-        self == Self::ROLE_0
+        self == Self::Role0
     }
 }
 
-impl Default for RoleIndex {
+impl Default for RoleId {
     fn default() -> Self {
-        Self::ROLE_0
+        Self::Role0
     }
 }
 
-impl From<u16> for RoleIndex {
-    fn from(value: u16) -> Self {
-        Self(value)
+impl TryFrom<u8> for RoleId {
```

Comment:
Use this instead: https://docs.rs/strum_macros/latest/strum_macros/derive.FromRepr.html
saves maintenance issues like adding a variant, and forgetting to update the conversion function.